### PR TITLE
Add a migration guide for WARNING for head/tail

### PR DESCRIPTION
**Pull request by Bodigrim**
Cf. #87.

### CODE REVIEW

File: guides/warning-for-head-and-tail.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+If you don't want to change any Haskell code and happy to carry on with `Data.List.{head,tail}`:
+
+1. If you use `-Werror`, set `-Wwarn=warnings-deprecations` to downgrade this area of compiler messages from errors to warnings, so that your build can succeed.
+
+2. If you are annoyed by warnings, set `-Wno-warnings-deprecations` to quash this area of compiler messages. You can also put `:set -Wno-warnings-deprecations` into `.ghci` config file.
+
+If you are happy to change code to eliminate warning messages:
+
+1. The easiest way to migrate away from `tail` is to replace it with `drop 1`. This is a drop-in replacement.
```
Comment (josephcsible) [inline guides/warning-for-head-and-tail.md:22]:
```suggestion
1. The easiest way to migrate away from `tail` is to replace it with `drop 1`. This is a drop-in replacement unless you are relying on `tail`'s partiality.
```

File: guides/warning-for-head-and-tail.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+# When
+
+At the very earliest, the change may appear in GHC 9.6 (~ Q1 2023), but more likely GHC 9.8 (~ Q4 2023).
+
+# How
+
+This is not a breaking change, unless a package forces `-Werror`.
+
+If you don't want to change any Haskell code and happy to carry on with `Data.List.{head,tail}`:
```
Comment (konsumlamm) [inline guides/warning-for-head-and-tail.md:14]:
```suggestion
If you don't want to change any Haskell code and want to carry on with using `Data.List.{head,tail}`:
```

File: guides/warning-for-head-and-tail.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+1. If you use `-Werror`, set `-Wwarn=warnings-deprecations` to downgrade this area of compiler messages from errors to warnings, so that your build can succeed.
+
+2. If you are annoyed by warnings, set `-Wno-warnings-deprecations` to quash this area of compiler messages. You can also put `:set -Wno-warnings-deprecations` into `.ghci` config file.
```
Comment (konsumlamm) [inline guides/warning-for-head-and-tail.md:18]:
```suggestion
2. If you are annoyed by warnings, set `-Wno-warnings-deprecations` to quash this area of compiler messages. You can also put `:set -Wno-warnings-deprecations` into `.ghci` config file.
   Note that this also disables any other custom warnings and deprecation warnings!
```

File: guides/warning-for-head-and-tail.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+# Related work
+
+Consider contributing to https://github.com/ghc-proposals/ghc-proposals/pull/454 and/or https://github.com/ghc-proposals/ghc-proposals/pull/541, which expand and make GHC warnings mechanism more flexible, allowing more mitigation strategies for this sort of changes.
+
+# PR template
+
+Here is a template, which you can use when raising PRs against affected libraries.
+
+> Title: Avoid Data.List.{head,tail}
```
Comment (konsumlamm) [inline guides/warning-for-head-and-tail.md:52]:
```suggestion
> Title: Avoid `Data.List.{head,tail}`
```

**Comment by Bodigrim** [comment]
Thanks @konsumlamm and @josephcsible!

File: guides/warning-for-head-and-tail.md [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    ```haskell
+    case foo of
+      [] -> error "<explain, why it is impossible to encounter an empty list here>"
+      hd : _ -> hd
+    ```
+
+3. Another option is to use [`Data.List.uncons`](https://hackage.haskell.org/package/base/docs/Data-List.html#v:uncons): replace any `head foo` with
+
+    ```haskell
+    fromMaybe (error "...") fst (uncons foo)
```
Comment (AndreasPK) [inline guides/warning-for-head-and-tail.md:36]: Doesn't this need to be something like below instead? As written this doesn't seem to typecheck.

```suggestion
    (fromMaybe (error "...") (fst <$> uncons foo))
```

Comment (Bodigrim) [reply guides/warning-for-head-and-tail.md]: Whoops, thanks, I meant `maybe` instead of `fromMaybe`.

**Comment by Bodigrim** [comment]
If there are no further suggestions, I'll merge this soon.

**Pull request merged**